docs/development/design/embed-api.rst

```diff
@@ -0,0 +1,154 @@
+Embed API
+=========
+
+The embedded API allows to embed content from docs pages in other sites.
+For a while it has been as an *experimental* feature without public documentation or real applications,
+but recently it has been used widely (mainly because we created a Sphinx extension).
+
+Due to this we need to have more friendly to use API,
+and general and stable enough to support it for a long time.
+
+.. contents::
+   :local:
+   :depth: 3
+
+Current implementation
+----------------------
+
+The current implementation of the API is partially documented in :doc:`/guides/embedding-content`.
+Some characteristics/problems are:
+
+- There are three ways of querying the API, and some rely on Sphinx's concepts like ``doc``.
+- Doesn't cache responses or doesn't purge the cache on build.
+- Doesn't support MkDocs.
+- It returns all sections from the current page.
+- Lookups are slow (~500 ms).
+- IDs returned aren't well formed (like empty IDs `#`).
+- The content is always an array of one element.
+- The section can be an identifier or any other four variants or the title of the section.
+- It doesn't return valid HTML for definition lists (``dd`` tags without a ``dt`` tag).
+- The client doesn't know if the page requires extra JS or CSS in order to make it work or look nice.
+
+Improvements
+------------
+
+These improvements aren't breaking changes, so we can implement them in the old and new API.
+
+- Support for MkDocs.
+- Always return a valid/well formed HTML block.
+
+New API
+-------
+
+The API would be split into two endpoints, and only have one way of querying the API.
+
+Get page
+--------
+
+Allow us to query information about a page, like its list of sections.
+
+.. http:get:: /_/api/v3/embed/pages?project=docs&version=latest&path=install.html
+
+   :query project: (required)
+   :query version: (required)
+   :query path: (required)
+
+   .. sourcecode:: json
+
+      {
+         "project": "docs",
+         "version": "latest",
+         "path": "install.html",
+         "title": "Installation Guide",
+         "url": "https://docs.readthedocs.io/en/latest/install.html",
+         "sections": [
+            {
+               "title": "Installation",
+               "id": "installation"
+            },
+            {
+               "title": "Examples",
+               "id": "examples"
+            }
+         ],
+         "extras": {
+            "js": ["https://docs.readthedocs.io/en/latest/index.js"],
+            "css": ["https://docs.readthedocs.io/en/latest/index.css"],
+         }
```

Comment:
Where does this data come from? Seems hard to parse it properly, and is providing a promise we can't meaningfully keep if we're expecting the CSS to work in a third-party page.

Reply:
I just expanded this in the new update

Reply:
The idea is good to me and could be useful in different cases. However, it's hard to make it work in _all of them_. For example, if the `.css` has relative URLs on its content it will break completely when loaded under a different path.

Reply:
Note that Sphinx 3.5 includes a feature to add JS and CSS files _only to the pages where "they are used"_ and this potentially kills our extension to render dynamic content. See https://github.com/sphinx-doc/sphinx/pull/8631#issuecomment-821155458

So, having a way to know what are the JS _and_ CSS files required in the target page is good.

Reply:
I was actually thinking to remove this, I put some of my thoughts in the document, basically most users want to inject content they know about or trust, so they know the required CSS/JS to make it work. I think you could refactor the extension to have a list of js/css files to include in all pages and auto-populate that with files from popular extensions based on some settings.

Reply:
I think we came down on "Let's remove it for now -- we can always add it later if we find a good use-case for it, but let's not start with it". Specifically, the _client_ should know enough about the content it's embedding to solve this usecase (eg. `if there is a sphinx-tabs classname, add the sphinx-tabs CSS`)

Reply:
> Specifically, the client should know enough about the content it's embedding to solve this usecase

:+1:

Also, dynamically adding CSS/JS is a security concern. I don't think there's any way to do that safely.